### PR TITLE
Added Rejected / Cancelled to History page

### LINKED ISSUE #1770: Rejected and canceled games should show up in history
Rejected and canceled games (and, when we have this, forfeited games) should show up in the history page; possibly with a way to include them or not, if we think (or hear) that people don't like that, but it should be *possible* to find them there.

### PR DESCRIPTION
**Pull request by blackshadowshade**
Fixes #1770.

http://jenkins.buttonweavers.com:8080/job/buttonmen-blackshadowshade/990/

### CODE REVIEW

**Comment by cgolubi1** [comment]
Okay.  At a glance, i'm wondering if this is going to make the whole history page wider, because Rejected/Cancelled is much longer than Active or Completed.  How strongly do we feel about differentiating?

**Comment by cgolubi1** [comment]
Anyway, there's nothing for me to do here except bring this up in virtualbox and take a look.

**Comment by cgolubi1** [comment]
Okay, so here's what it looks like when i do a history search on myself (note that it's scrolled to the right, because the History page is too wide for my browser, but that's not really new):
history_search

**Comment by cgolubi1** [comment]
The thing that i think is definitely wrong is boldfacing Rejected/Cancelled - that column seems to have "Completed" boldfaced and "In Progress" in italics - my suggestion would be to have no font formatting on "Rejected/Cancelled".

I don't want to get people used to the idea of thinking of Rejected and Cancelled as separate *game* states --- they're not, they're actions which we've given separate names, but they lead to the same game state of a game which was thrown away before it was started.  IMO encouraging people to think of a rejected game as different from a cancelled game takes up more horizontal space for the benefit of making people's expectations slightly more wrong.  :>)  So i'd rather we not go there more than needed, and thus would favor simply calling this state "Rejected".

Otherwise, this appears to work.

File: src/engine/BMInterface.php [hunk 1]
```diff
@@ -1834,7 +1834,9 @@ protected function apply_all_filters($searchFilters, &$where, &$whereParameters)
             $whereParameters[':status_%%%'] = $searchFilters['status'];
         } else {
             // We'll only display games that have actually started
```
Comment (AdmiralJota) [inline src/engine/BMInterface.php]: Does this comment still strictly apply?

**Comment by AdmiralJota** [comment]
I think that either "Rejected" or "Cancelled" would work by itself. It's unfortunate that the word "Aborted" has other undesirable connotations, since I think it'd be exactly right otherwise. Other possibilities might be "Unaccepted", "Annulled", "Retracted", "Called Off", "Void", "Forsaken"...

As for formatting, I agree that it shouldn't be bolded or italicized. But it might be reasonable to make the color some shade of grey rather than black.

Also, as @irilyth mentioned in #1770, I think we should still have a way to show all the games that have actually been started (i.e., all active and completed games). I can see two approaches for that, both with pros and cons:

* Keep the blank, default option to ACTIVE and COMPLETED, and add an extra option at the end of the dropdown for "All (including Rejected)". On the one hand, I think the current behavior of the default being all games <i>that have actually started at some point</i> is probably what most people would expect and be looking for as the default. But on the other hand, it means that it's now not as obvious what the default option means without making inferences from the rest of the list.

* Have the blank, default option be all games (including rejected ones), and add an item to the list after Active and Completed that's something like "Active / Completed" or "All Started Games" or "This Is The Option You Probably Mean to Pick Unless You Really Want To See Rejected Games Mixed In With Everything Else For Some Reason". This has the upside that it's clear what every option means (provided you realize that rejected games are a thing that can appear in the history list at all), but the downside that it seems to make the dropdown options kinda more awkward. And it includes Rejected games in the default case, whereas I strongly suspect that wanting to include rejected games is going to be a special rare use case rather than the standard one.

* I suppose that, technically, there's the third option of "change the dropdown into some wacky multi-select control and let users check off individual statuses", but A) that seems like a much bigger change than we should be considering right now, and B) there are additional implementation implications and UI decisions that would need to follow from that. Although it might be something we may want to consider at some point in the future.

So, while we're here, adding Rejected games to the history list... why are we excluding challenges that haven't been accepted or declined yet? It seems like they both have about equal standing in the "is it really a game?" question -- arguably more standing, since they still show up in the active games list and can still be promoted to real games.

**Comment by cgolubi1** [comment]
If we're suggesting replacement words entirely, i'd add Discarded and Abandoned as suggestions.

**Comment by cgolubi1** [comment]
> Also, as @irilyth mentioned in #1770, I think we should still have a way to show all the games that have actually been started (i.e., all active and completed games)

This seems like a fine idea, but IMO this pull request would still be an improvement without it.

**Comment by cgolubi1** [comment]
> So, while we're here, adding Rejected games to the history list... why are we excluding challenges that haven't been accepted or declined yet?

This is a fair question - we could probably add "New" games to the pulldown pretty easily using the same logic we already have.

**Comment by irilyth** [comment]
> I don't want to get people used to the idea of thinking of Rejected and Cancelled as separate game states --- they're not, they're actions which we've given separate names, but they lead to the same game state of a game which was thrown away before it was started.

One way in which they're different from a player's point of view is that you might be looking for games that I created but that other people rejected, and not care about games that I created and then canceled.

I suppose there isn't an obvious way in the history interface to differentiate between games I created and the other player rejected, and games another player created and I rejected, is there? (There is in the DB, though, right? In the former, I'm "player" and in the latter I'm "opponent"?)

I wonder if we want to start thinking about the current history page as the "simple game search" page, and think about an "advanced game search" page where you could search by criteria that you can't currently enter on the current page (like creator, description, was this a followup game, etc).

**Comment by blackshadowshade** [comment]
Right, I've changed the word to 'Cancelled' and grayed it out.

http://jenkins.buttonweavers.com:8080/job/buttonmen-blackshadowshade/994/ (if it passes)

**Comment by cgolubi1** [comment]
Okay, let me check this out and see if it addresses my concerns - there's still Jota's request to be able to see New games in the search, but IMO we can/should make that a separate issue.

**Comment by cgolubi1** [comment]
I opened #1839 and #1840 for things which are more about New games on the History page than about Cancelled games on the history page.  Now i am going to merge this.

**Pull request merged**